src/main/scala/latis/server/DataService.scala

```diff
@@ -1,15 +1,106 @@
+import org.http4s.circe._
 import org.http4s.dsl.Http4sDsl
+import org.http4s.dsl.io._
 
 /** Implements the `/data` endpoint. */
 class DataService[F[_]: Effect] extends Http4sDsl[F] {
+  import QueryDecoders._
 
   val service: HttpService[F] =
     HttpService[F] {
-      case GET -> Root / "data" =>
+      case GET -> Root / "data"
+          :? IdMatcher(_)
+          +& MinTimeMatcher(_)
+          +& MaxTimeMatcher(_)
+          +& ParamMatcher(_)
+          +& IncludeMatcher(_)
+          +& FormatMatcher(_) =>
         Ok("Hello from HAPI!")
+      // Return a 1400 error if the required parameters are not given.
+      case GET -> Root / "data" :? _ =>
+        BadRequest(Status.`1400`.asJson)
+    }
+}
+
+/**
+ * Represents a request for data.
+ *
+ * @param id HAPI dataset ID
+ * @param minTime lower time bound (inclusive)
+ * @param maxTime upper time bound (exclusive)
+ * @param parameters list of parameter names to return
+ * @param header whether to include the metadata header
+ * @param format output format
+ */
+final case class DataRequest(
+  id: String,
+  minTime: String,
+  maxTime: String,
+  parameters: Option[List[String]],
+  header: Boolean,
+  format: String
+)
+
+/** Wrapper for `include` parameter. */
+final case class Include(header: Boolean) extends AnyVal
+
+object Include {
+
+  implicit val includeDecoder: QueryParamDecoder[Include] =
+    new QueryParamDecoder[Include] {
+      override def decode(qpv: QueryParameterValue) = qpv.value match {
+        case "header" => Include(true).validNel
+        case _ =>
+          ParseFailure(
+            "Invalid value for 'include' parameter",
+            s"Invalid value for 'include' parameter: ${qpv.value}"
+          ).invalidNel
+      }
+    }
+}
+
+/** Wrapper for `format` parameter. */
+final case class Format(format: String) extends AnyVal
+
+object Format {
+
+  implicit val formatDecoder: QueryParamDecoder[Format] =
+    new QueryParamDecoder[Format] {
+      override def decode(qpv: QueryParameterValue) = qpv.value match {
+        case fmt @ ("csv" | "binary" | "json") => Format(fmt).validNel
+        case _ =>
+          ParseFailure(
+            "Invalid value for 'format' parameter",
+            s"Invalid value for 'format' parameter: ${qpv.value}"
+          ).invalidNel
+      }
+    }
+}
+
+object QueryDecoders {
+
+  object IdMatcher extends QueryParamDecoderMatcher[String]("id")
+  object MinTimeMatcher extends QueryParamDecoderMatcher[String]("time.min")
+  object MaxTimeMatcher extends QueryParamDecoderMatcher[String]("time.max")
+  object ParamMatcher extends OptionalQueryParamDecoderMatcher[List[String]]("parameters")
+  object IncludeMatcher extends OptionalQueryParamDecoderMatcher[Include]("include")
+  object FormatMatcher extends OptionalQueryParamDecoderMatcher[Format]("format")
+
+  /** Decoder for simple CSV query parameters. */
+  implicit def csvDecoder[A : QueryParamDecoder]: QueryParamDecoder[List[A]] =
```

Comment:
Empty parameter lists (i.e. `parameters=` in a query) are supposed to return an error.

I'll take a look at this and see how hard it would be to make that happen. If I can't figure it out quickly, I'll just make a new ticket to handle it later.

Reply:
I've changed this so those queries are rejected.